### PR TITLE
Fix conversion from/to UTF-32

### PR DESCRIPTION
**Pull request by gracicot**
Fixes the issue #8954 and should fix the tests for #8931 

I made it work by copying the code that translated UTF-8, but using the UTF 16 and 32 functions.

If it smells to much like duplicated code, I can try to improve it.

### CODE REVIEW

**Comment by kripken** [comment]
Thanks for the PR @gracicot!

@jgravelle-google can you please take a look?

**Comment by gracicot** [comment]
@sbc100 Was there something wrong with this patch? Is there something I can improve? Why was this closed?

**Comment by sbc100** [comment]
The deletion of the `incoming` branch triggered the automatic closing of this branch.  My apologies.  This was an unexpected side effect.

Re-opening and re-targeting to master.

File: src/embind/embind.js [hunk 1]
```diff
@@ -729,39 +729,79 @@ var LibraryEmbind = {
...
-            for (var i = 0; i < length; ++i) {
-                HEAP[start + i] = value.charCodeAt(i);
+            if (value instanceof ArrayBuffer) {
+                value = new Uint8Array(value);
+            }
+
+            var valueIsOfTypeString = (typeof value === 'string');
+
+            if (!(valueIsOfTypeString || value instanceof Uint8Array || value instanceof Uint8ClampedArray || value instanceof Int8Array)) {
+                throwBindingError('Cannot pass non-string to std::basic_string');
```
Comment (jgravelle-google) [inline src/embind/embind.js]: This should say `std::wstring`?

File: src/embind/embind.js [hunk 2]
```diff
@@ -729,39 +729,79 @@ var LibraryEmbind = {
...
         name: name,
         'fromWireType': function(value) {
-            var HEAP = getHeap();
             var length = HEAPU32[value >> 2];
-            var a = new Array(length);
-            var start = (value + 4) >> shift;
-            for (var i = 0; i < length; ++i) {
-                a[i] = String.fromCharCode(HEAP[start + i]);
+            var HEAP = getHeap();
+            var str;
```
Comment (jgravelle-google) [inline src/embind/embind.js]: I'm relatively ok with this duplication, but would appreciate a comment inline that this is copied from register_std_string's registerType

File: src/embind/embind.js (hunk 1 above)
Comment (gracicot) [reply src/embind/embind.js]: Well, since it could be `std::wstring`, `std::u16string` or `std::u32string` I stayed general by saying `std::basic_string`

Comment (gracicot) [reply src/embind/embind.js]: Would that:

> Cannot pass non-string to `std::wstring`, `std::u16string` or `std::u32string`

Be better?

Comment (jgravelle-google) [reply src/embind/embind.js]: To *me specifically* that sounds better, but I didn't know that basic_string was the base of wstring et al, so it looked like a copy-paste error. https://en.cppreference.com/w/cpp/string/basic_string lists all those, so maybe basic_string is fine.

Having said that, I imagine this code breaks if you pass any basic_string that isn't wstring u16string or u32string, so probably better to be explicit about what we support.

Comment (gracicot) [reply src/embind/embind.js]: @jgravelle-google I just realized that the name is sent as parameter to the function. Now it's going to print the right type name everytime, even when merging the other patch.

**Comment by gracicot** [comment]
I made one last change to remove support for `UInt8Array` and friends. This only applied with ASCII `std::string`. It has nothing to do with unicode string and I left it there as a copy-paste error.

**Pull request merged**